File: src/main/java/leetcode/Problem81.java
package leetcode;

@Algorithms(Algorithm.BINARY_SEARCH)
@BeatsPercent(100)
@TimeComplexity(average = Complexity.LOGARITHMIC_N, worst = Complexity.LINEAR_N)
@SpaceComplexity(worst = Complexity.CONSTANT)
public class Problem81 {
    public boolean search(int[] nums, int target) {
        int lo = 0;
        int hi = nums.length - 1;
        while (lo < hi) {
            int mid = lo + (hi - lo)/2;
            if (target == nums[mid]) {
                return true;
            }
            if (nums[lo] < nums[mid]) { // left part is sorted
                if (target < nums[mid] && target >= nums[lo]) {
                    hi = mid - 1;
                } else {
                    lo = mid + 1;
                }
            } else if (nums[mid] < nums[lo]) { // right part is sorted
                if (target > nums[mid] && target <= nums[hi]) {
                    lo = mid + 1;
                } else {
                    hi = mid - 1;
                }
            } else { // no clue where it is, but it's not nums[lo], because nums[lo] == nums[mid]
                lo++;
            }
        }
        return nums[lo] == target;
    }
}

File: src/problem_81.py
class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        lo, hi = 0, len(nums) - 1
        while lo < hi:
            mi = (lo + hi) // 2
            if target == nums[mi]:
                return True
            if nums[lo] == nums[mi]:
                lo += 1
            elif nums[lo] <= target < nums[mi] or (nums[mi] <= nums[hi] and (target < nums[mi] or target > nums[hi])):
                hi = mi - 1
            else:
                lo = mi + 1
        return nums[lo] == target
